tests/exercise_instructions.rs

```diff
@@ -0,0 +1,506 @@
+                    addr as i32,
+                    addr >> 32,
+                ),
+                &mut prng,
+                25,
+            );
+
+            let addr = rand % 73 + 0x4_0000_0000i64 - offset as i64;
+            test_ins(
+                format!(
+                    "mov32 r{tmp},{}
+                    mov64 r{src},{:#x}
+                    lsh64 r{src},32
+                    or64 r{src},r{tmp}
+                    stdw [r{src}{offset:+}], {imm}",
+                    addr as i32,
+                    addr >> 32,
+                ),
+                &mut prng,
+                25,
+            );
+
+            // unconditional jump
+            test_ins(
+                format!(
+                    "ja 1
+                     xor64 r{src},0x12345678
+            1:",
+                ),
+                &mut prng,
+                21,
+            );
+
+            for jc in [
+                "jeq", "jgt", "jge", "jlt", "jle", "jset", "jne", "jsgt", "jsge", "jslt", "jsle",
+            ] {
+                test_ins(
+                    format!(
+                        "{jc} r{src}, {imm}, l1
+                        or64 r{src},0x12345678
+                        ja l2
+                    l1:
+                        and64 r{src},0x12345678
+                        ja l2
+                    l2:",
+                    ),
+                    &mut prng,
+                    23,
+                );
+            }
+
+            imm &= 63;
+
+            test_ins(format!("lsh64 r{src}, {imm}"), &mut prng, 21);
+            test_ins(format!("rsh64 r{src}, {imm}"), &mut prng, 21);
+            test_ins(format!("arsh64 r{src}, {imm}"), &mut prng, 21);
+            test_ins(format!("hor64 r{src}, {imm}"), &mut prng, 21);
+
+            imm &= 31;
+
+            test_ins(format!("lsh32 r{src}, {imm}"), &mut prng, 21);
+            test_ins(format!("rsh32 r{src}, {imm}"), &mut prng, 21);
+            test_ins(format!("arsh32 r{src}, {imm}"), &mut prng, 21);
+
+            test_ins(format!("be64 r{src}"), &mut prng, 21);
+            test_ins(format!("be32 r{src}"), &mut prng, 21);
+            test_ins(format!("be16 r{src}"), &mut prng, 21);
+        }
+    }
+}
+
+fn test_ins(ins: String, prng: &mut SmallRng, cu: u64) {
+    let mut input = [0u8; 80];
+
+    prng.fill_bytes(&mut input);
+
+    let asm = format!(
+        "
+        ldxdw r9, [r1+72]
+        ldxdw r8, [r1+64]
+        ldxdw r7, [r1+56]
+        ldxdw r6, [r1+48]
+        ldxdw r5, [r1+40]
+        ldxdw r4, [r1+32]
+        ldxdw r3, [r1+24]
+        ldxdw r2, [r1+16]
+        ldxdw r0, [r1+0]
+        ldxdw r1, [r1+8]
+        {ins}
+        xor64 r0, r1
+        xor64 r0, r2
+        xor64 r0, r3
+        xor64 r0, r4
+        xor64 r0, r5
+        xor64 r0, r6
+        xor64 r0, r7
+        xor64 r0, r8
+        xor64 r0, r9
+        exit"
+    );
+
+    test_interpreter_and_jit_asm!(asm.as_str(), input, (), TestContextObject::new(cu));
```

Comment:
Are we testing `call` and `exit`?

Reply:
That's true, however testing those is a bit trivial/uninteresting